### ISSUE TITLE
[WIP] Tech Spec: Searching

### ISSUE THREAD

**Issue by mdbenjam** [department-of-veterans-affairs/caseflow#3469]
## Context
Searching text within documents has been a painful process for users. The fundamental problem is that by relying on the browser's ctrl + f search functionality we require all the text to be in the DOM. We've gone through three iterations to make it better.

First, we drew only the pages that needed to be rendered. However, this made searching impossible for users since the full text was not on the page. Second, we drew all the pages all the time. However, this made navigating around the doc slow, and searching wouldn't uncover all of the results until the entire PDF had loaded (for large docs, a long time). Third, we drew text for all the pages, but only drew images for pages within view. While better than the first two, drawing all the text for a 3000 page doc still caused a big slow down.

While testing with users we learned two things:
1) Users expect to be able to search text quickly.
1) Users expect to be able to move within the document smoothly.

## Overview
In order to avoid drawing all the pages all the time we need to write our own search function. We will emulate the browser's search in javascript. Typing a term will give a result count, clicking next and previous (or enter and shift+enter) will move between search hits. Each search hit will be highlighted within the document.

Finding every search result requires combing through a lot of text. We want to do this in an efficient way so we propose a tiered searching structure. We'll use redux-search to narrow down the results to the set of pages with that word/phrase. Then we'll look at each page individually to linearly count the number of instances of that text. This will give us the total number of hits. Once we have the total number of hits we can highlight the text, and move between hits by keeping track of the index of the search.

### [Redux-Search](https://github.com/bvaughn/redux-search)
Redux search is a library that indexes objects based on their text fields for easy searching. We can insert an object for every page that has the page text as a field. Then we can easily run an efficient search against all the pages.

### Linear Search
For every page that has a search result we can use a [regex to do a linear search for the text](https://stackoverflow.com/questions/4009756/how-to-count-string-occurrence-in-string). In the worst case this could require us to linearly search the entire document, but should normally limit us to just the set of pages with matches. We will memoize these results in a selector.

### Highlighting Results
We'll use [mark.js](https://markjs.io/) to highlight search results. Mark.js traverses the DOM looking for the search text to highlight. It also provides callbacks to get references of the DOM elements it highlights. We will store the marked DOM elements in the PdfPage components.

### Moving Between Search Results
To move between search results we'll maintain an index in Redux that records which search result we're currently highlighting. As this value is incremented we let the PdfPage components determine which page has the hit. That component is then responsible for changing the highlighting of the focused hit, as well as moving the scroll window so that it's visible.

## Next Steps
To speed up the document rendering we need to stop rendering the text of every page. We can start by turning off text rendering until the page is in view. Long term we'll explore using [react-virtualized](https://github.com/bvaughn/react-virtualized) so that we don't hand roll our own windowing code.

## Rollout
1) Add a search field hidden behind a feature flag. This search field will use redux-search to find the pages with the search term.
1) Add in linear search to count total number of instances of the search term.
1) Do performance testing. Record how long it takes to find all the search terms.
1) Do performance testing on a similar setup without redux search (this should be easily modifiable. Just have redux search return that the search term is on every page.) Record how long it takes to find all the search terms. Compare these values to verify adding redux-search is worth it.
1) If performance is acceptable we continue by implementing highlighting and navigating between search terms.

## Open Questions
- What is the minimum number of characters typed before we start searching. It seems like 1 is too few and will have too many hits. Maybe 2, or 3?
- How frequently does the linear search hit every page?
- How do that regex method and Mark.js deal with searching "aba" in "ababa". Is it one or two matches, and is it consistent?
   - We should consume letters as we search them. i.e. in the above example "ababa" would match once for the first "aba"

**Comment by nickheiner-usds** [comment]
Nice tech spec. I think that it would generally be easier to submit them as a PR to a markdown file, so we can comment on each line and track changes over time. (No need to change it now.)

I think this generally seems like it'll work.

For the built-in browser search, this is the "aba" v. "ababa" solution:

![image](https://user-images.githubusercontent.com/25331532/31507002-27e0cd2c-af47-11e7-8a62-1a0273853309.png)

It makes sense to me to say that the first match consumes the characters, and subsequent match searching starts after the end of the first match.

My biggest open question: how will this relate to built-in browser search from a UI perspective? Will we capture the `control/cmd-f` event and not let the user do browser search? Is that a losing game, because we can't stop the user from using the browser menus to search? Are we going to have confusing situations where people try to search with the built-in browser search and are confused when it doesn't work?

**Comment by mdbenjam** [comment]
I'll convert this to a PR later. I think it's a good idea to have it checked in.

I like the idea that searching consumes the first instances of those characters.

> My biggest open question: how will this relate to built-in browser search from a UI perspective?

That's a great question for designers. I sort of sprung this on them, so I don't think @gnakm has had time think too much about the UI. It's probably worth hijacking ctrl + f. I don't really want to hijack browser functionality... but in this instance I think it's our best option. Yes people can get around it through the menus. But I think if we make a clear UI about where to search most people will either click on the search bar or ctrl + f to get there.

**Comment by evankroske** [comment]
Is there a way to simplify our searching approach? This tech spec involves introducing two libraries, indexes, and caches. Could we start with a simpler, slower approach that we could iterate on?

Would document searching be useful to other apps in the VA? Would it make more sense to add a search API to eFolder express or a new service?

**Comment by mdbenjam** [comment]
> Is there a way to simplify our searching approach?

Possibly, but I did a lot of experimentation before getting to this point. I first tried just to do linear searching, and my results are pretty negative. The searching took a long time, it all also happened on the main thread causing noticeable slow downs. [Redux-search](https://github.com/bvaughn/redux-search/tree/master/docs#searchapi) uses web workers, when it can, to keep our UI responsive.

> This tech spec involves introducing two libraries, indexes, and caches.

I don't see a good way to avoid using mark.js. We don't control rendering of the text. PDFJS does it for us. We could try to render the text in React, but I looked into the code that does the text rendering in PDFJS and it is incredibly complicated. We'd likely have to duplicate it. By using mark.js we can highlight terms without having to render them ourselves.

We could get by without redux-search, but it would both make our search slower and more complicated to write since we'd have to hand-roll our own solution.

> Would document searching be useful to other apps in the VA? Would it make more sense to add a search API to eFolder express or a new service?

We have not needed this functionality yet, and I don't foresee our other products needing it. For example I doubt work queue will require arbitrary text searching. I think it's worthwhile to start with this, then expand to an API if we really need it. Moreover, a large part of this is finding where in the document the words are and allowing users to move between them. That seems like a pretty Reader-specific task.

**Comment by sunil-sadasivan** [comment]
> My biggest open question: how will this relate to built-in browser search from a UI perspective?

UX is def not my strong suite, however I notice a lot of apps override the default browser ctrl+f functionality, so I think this feels acceptable for us to do this in the reader app. 
 (google docs)
<img width="1158" alt="screenshot 2017-10-12 19 58 01" src="https://user-images.githubusercontent.com/31223104/31524440-acd1523c-af87-11e7-84f1-21cf099eb196.png">

I'm not too concerned with bringing in two new libraries into the project, however (and it feels like Mark's well aware of this) we're doing a lot here to build this functionality.  That feels to be the general theme of this project (implementing a reader from scratch vs using an existing solution), so I'm not too surprised we have to implement search functionality like this too.  

@mdbenjam, I think my main concern aligns with Evan in that it's important to ensure we can find ways to validate assumptions quickly.  I'm generally always hesitant with such a large implementation change set vs a more exploratory approach.  Is there a way to break the implementation into discrete phases?  This plan generally feels good to me. Having the mindset of breaking this out into implementation chunks in which we can validate the success and performance of the approach would be awesome.

**Comment by mdbenjam** [comment]
Added a rollout plan, so we can validate assumptions around search speed before moving on to the UI parts.

**Comment by nickheiner-usds** [comment]
Good instinct @evankroske and @sunil-sadasivan to point out that this sounds like a complicated solution to the problem. I do agree with @mdbenjam that we've already done some exploratory / simple solutions and have not been satisfied with the result. :Smile:

**Comment by nickheiner-usds** [comment]
For performance testing, this could be a good place to use Google Analytics Page Speed Timings. We can record arbitrary times and send them to GA, and then we'll have real-world usage.

**Comment by mdbenjam** [comment]
Closing this in favor of this PR: https://github.com/department-of-veterans-affairs/appeals-pm/pull/1608
